crates/runtime/src/model/tool_use_responses.rs

```diff
@@ -364,6 +417,246 @@ impl Responses for ToolUsingResponses {
     }
 }
 
+struct CustomResponseStream {
+    receiver: mpsc::Receiver<Result<ResponseEvent, OpenAIError>>,
+}
+
+impl Stream for CustomResponseStream {
+    type Item = Result<ResponseEvent, OpenAIError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.receiver.poll_recv(cx)
+    }
+}
+
+#[allow(clippy::too_many_lines)]
+fn make_responses_stream(
+    span: Span,
+    request_context: Arc<RequestContext>,
+    model: ToolUsingResponses,
+    req: CreateResponse,
+    mut s: ResponseStream,
+) -> ResponseStream {
+    let (sender, receiver) = mpsc::channel(100);
+    let sender_clone = sender.clone();
+
+    tokio::spawn(
+        request_context
+            .scope(async move {
+                let function_call_states: Arc<Mutex<HashMap<String, FunctionCall>>> =
+                    Arc::new(Mutex::new(HashMap::new()));
+                let completed_function_calls: Arc<Mutex<Vec<FunctionCall>>> =
```

Comment:
This isn't really completed function calls (as in, we've called the function and gotten its results). I would call this "ready_to_call_functions"